File: examples/factors_sdk/typescript/02_resolve_with_explain.ts
/**
 * 02 — Resolve an activity with full explain payload (Pro+ tier).
 *
 * Shows the 7-step cascade result, assumptions, gas breakdown, and
 * quality score for a concrete activity.
 */
import { FactorsClient, FactorsAPIError, TierError } from '@greenlang/factors';

async function main(): Promise<void> {
  const client = new FactorsClient({
    baseUrl: process.env.GL_FACTORS_API_URL ?? 'https://api.greenlang.io',
    apiKey: process.env.GL_FACTORS_API_KEY,
    edition: process.env.GL_FACTORS_EDITION ?? 'ef_2026_q1',
    methodProfile: 'corporate_scope1',
  });

  try {
    const resolved = await client.resolve(
      {
        activity: 'diesel combustion',
        method_profile: 'corporate_scope1',
        jurisdiction: 'US-CA',
        reporting_date: '2026-01-15',
        facility_id: 'plant-42',
      },
      { alternates: 3 },
    );

    console.log('Chosen factor :', resolved.chosen_factor_id ?? resolved.factor_id);
    console.log('Cascade step  :', resolved.fallback_rank, resolved.step_label);
    console.log('Method profile:', resolved.method_profile);
    console.log('Why           :', resolved.why_chosen);
    console.log('DQS           :', resolved.quality_score?.overall_score);
    console.log('CO2e basis    :', resolved.co2e_basis);
    console.log('Gas breakdown :', resolved.gas_breakdown);
    console.log('Assumptions   :');
    for (const a of resolved.assumptions ?? []) console.log('  -', a);
  } catch (err) {
    if (err instanceof TierError) {
      console.error('This endpoint requires Pro+ tier:', err.message);
      console.error('Remediation:', err.remediation);
    } else if (err instanceof FactorsAPIError) {
      console.error(`API error (HTTP ${err.statusCode}):`, err.message);
    } else {
      throw err;
    }
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: examples/factors_sdk/python/02_resolve_with_explain.py
# -*- coding: utf-8 -*-
"""Example 02 — Full resolution + explain cascade.

Submits a ResolutionRequest and prints the chosen factor with its
7-step derivation chain + alternates.  Requires Pro+ tier.
"""
from __future__ import annotations

import json
import os

from greenlang.factors.sdk.python import FactorsClient
from greenlang.factors.sdk.python.models import ResolutionRequest


def main() -> None:
    base_url = os.environ.get("GREENLANG_FACTORS_BASE_URL", "http://localhost:8000")
    api_key = os.environ.get("GREENLANG_FACTORS_API_KEY")

    request = ResolutionRequest(
        activity="diesel combustion stationary",
        method_profile="corporate_scope1",
        jurisdiction="US",
        reporting_date="2026-06-01",
    )

    with FactorsClient(base_url=base_url, api_key=api_key) as client:
        resolved = client.resolve(request, alternates=5)

    print("Chosen factor:", resolved.chosen_factor_id)
    print("Fallback rank:", resolved.fallback_rank, "/", 7)
    print("Step label:  ", resolved.step_label)
    print("Why chosen:  ", resolved.why_chosen)
    print("Method:      ", resolved.method_profile)
    print("Edition:     ", resolved.edition_id)

    if resolved.quality_score:
        print(f"Quality:     {resolved.quality_score.overall_score} ({resolved.quality_score.rating})")

    if resolved.gas_breakdown:
        gb = resolved.gas_breakdown
        print(
            f"Gas breakdown: CO2={gb.CO2} CH4={gb.CH4} N2O={gb.N2O} "
            f"(CH4_gwp={gb.ch4_gwp}, N2O_gwp={gb.n2o_gwp})"
        )

    print(f"\nAlternates considered ({len(resolved.alternates)}):")
    for a in resolved.alternates:
        print("  -", a.get("factor_id"), "score=", a.get("score"))

    if resolved.assumptions:
        print("\nAssumptions:")
        for note in resolved.assumptions:
            print("  *", note)

    print("\nRaw explain payload:")
    print(json.dumps(resolved.explain, indent=2, default=str))


if __name__ == "__main__":
    main()
